workflow/v1_2/workflow.go

```
@@ -78,21 +78,99 @@ type bakedInFile struct {
+func maybeMigrateFileMap(fm *effects.FileManager, param json.RawMessage) (bool, json.RawMessage, error) {
+	bifMap := make(map[string]*bakedInFile, 0)
+	if err := json.Unmarshal([]byte(param), &bifMap); err != nil {
+		return false, nil, nil
+	}
+
+	js := make(map[string]wtype.File, len(bifMap))
+	for k, bif := range bifMap {
+		if f, err := moveDataToFile(fm, bif); err != nil {
+			return false, nil, err
+		} else if f == nil {
+			return false, nil, nil
+		} else {
+			js[k] = *f
+		}
+	}
+
+	if msg, err := json.Marshal(js); err != nil {
+		return false, nil, err
 	} else {
-		if f, err := fm.WriteAll(bif.Bytes.Bytes, bif.Name); err != nil {
-			return nil, err
-		} else if fbs, err := json.Marshal(f.AsInput()); err != nil {
-			return nil, err
+		return true, msg, nil
+	}
+}
+
+// maybeMigrateFileArray returns bool indicating whether this could be serialised as an array of file references, then either the original json or transformed json if successful.
+func maybeMigrateFileArray(fm *effects.FileManager, param json.RawMessage) (bool, json.RawMessage, error) {
+	bifArr := make([]*bakedInFile, 0)
+
+	if err := json.Unmarshal([]byte(param), &bifArr); err != nil {
+		return false, nil, nil
+	}
+
+	js := make([]wtype.File, len(bifArr))
+	for i, bif := range bifArr {
+		if f, err := moveDataToFile(fm, bif); err != nil {
+			return false, nil, err
+		} else if f == nil {
+			return false, nil, nil
 		} else {
-			return json.RawMessage(fbs), nil
+			js[i] = *f
 		}
 	}
+
+	if msg, err := json.Marshal(js); err != nil {
+		return false, nil, err
+	} else {
+		return true, msg, nil
+	}
+}
+
+// maybeMigrateFileFlat returns bool indicating whether this could be serialised as a file reference, then either the original json or transformed json if successful.
+func maybeMigrateFileFlat(fm *effects.FileManager, param json.RawMessage) (bool, json.RawMessage, error) {
```

Ok, so this comment applies to all three of maybeMigrateFile* functions:

* the bool and the error can be combined. I 100% agree to ignore the error in the case that the initial unmarshal fails.
* but from there on, it should always be the case that err != nil implies not success.
* in all 3 functions, there's a situation where you check f == nil, after checking the err != nil. That *should* be impossible - I would arrange for the contract of moveDatatoFile to be that if err == nil then the file is non nil.
* finally, by reducing the return type to (json.RawMessage, error) I think you can then just `return json.Marshal(f)` as the final step, which would make all of these shorter still.